HDDS-10108. [hsync] Adopt RATIS-1994 to reduce hsync ops latency

smengcl opened the pull request:
## What changes were proposed in this pull request?

Adopt the new Ratis API added in RATIS-1994 to (hopefully) reduce hsync latency.

Basically the idea is that ratis client `send()` call alone would suffice. And we could ditch `watchForCommit` in hsync (writeChunk + putBlock) path.

Requires Ratis 3.1.0.

Note:

1. In order for `failedServers` list to be properly appended when a datanode lags behind in Ratis server `watch()` (in which case server throws `NotReplicatedException`), as checked by `TestFailureHandlingByClient.testDatanodeExclusionWithMajorityCommit`, significant adjustments has been made around `sendRequest()` and `sendCommand()`.
2. Datanodes `raft.server.watch.timeout` config has been lowered from [180s](https://github.com/apache/ozone/blob/f3a0dbd61af864369ee74b565f82d714e9e37d38/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/conf/DatanodeRatisServerConfig.java#L56-L64) to 10s ([Ratis default](https://github.com/apache/ratis/blame/dac27e4fd3df042fcf50ed549404e4d82708ed2b/ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java#L313-L315)), as it has to be lower than `raft.client.rpc.request.timeout` ([currently 60s in Ozone](https://github.com/apache/ozone/blob/f3a0dbd61af864369ee74b565f82d714e9e37d38/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/ratis/conf/RatisClientConfig.java#L64-L72)) for `NotReplicatedException` to be propagated

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10108

## How was this patch tested?

- Existing tests must all pass. Manual latency test pending.

jojochuang commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java:
```diff
@@ -589,7 +589,7 @@ private void handleFlushInternal(boolean close)
     }
     waitOnFlushFutures();
-    watchForCommit(false);
+//    watchForCommit(false);
```
watchForCommit is used by waitForFlushAndCommit too.

RatisBlockOutputStream.hsync() calls waitForFlushAndCommit() call, which is redundant if handleFlush() is already called. Perhaps we can update it like:

      if (getBufferPool() != null && getBufferPool().getSize() > 0) {
        handleFlush(false);
      } else {
        waitForFlushAndCommit(false);
      }

smengcl replied:
You are right. It looks redundant. Done

jojochuang reviewed (commented):
Looks good. Just one comment:

szetszwo commented:
@smengcl , thanks for working on this!

`TestBlockOutputStreamCorrectness` can pass if we override `watchForCommit(..)` in `CommitWatcher`; see https://issues.apache.org/jira/secure/attachment/13066441/6014_idea.patch (Need to update `TestCommitWatcher.testReleaseBuffersOnException`)

smengcl commented on hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockOutputStreamCorrectness.java:
```diff
@@ -56,37 +57,28 @@
-  @Test
-  public void test() throws IOException {
+  @ParameterizedTest
+  @ValueSource(ints = { 1, 1024, 1024 * 1024 })
```
Reverted HDDS-10173 from this PR.

smengcl commented:
> @smengcl , thanks for working on this!
> 
> `TestBlockOutputStreamCorrectness` can pass if we override `watchForCommit(..)` in `CommitWatcher`; see [issues.apache.org/jira/secure/attachment/13066441/6014_idea.patch](https://issues.apache.org/jira/secure/attachment/13066441/6014_idea.patch) (Need to update `TestCommitWatcher.testReleaseBuffersOnException`)

Thanks @szetszwo for the help. It works!

smengcl commented:
I noticed some refactoring (HDDS-10173) done in `TestBlockOutputStreamCorrectness` in `master` but not in the feature branch so I didn't change it at the moment to reduce merge conflict (test would still pass, the key is `CommitWatcher`).

smengcl commented:
3 test failures remaining at this point, not including flaky ones:

1. [`TestRandomKeyGenerator.testKeyLargerThan2GB`](https://github.com/apache/ozone/actions/runs/8907392352/job/24479784720?pr=6014#step:6:2100) -- somehow key write always fails with this large 2.01 GB key in this test. It can be reproduced locally. If I change the key size to 0.01 GB it passes. I suspect this has something to do with client write buffer management as well. But haven't quite figured out why.
2. [`TestFailureHandlingByClient.testDatanodeExclusionWithMajorityCommit`](https://github.com/apache/ozone/actions/runs/8907392352/job/24479780977?pr=6014#step:6:2151) -- it fails because the test is expecting the datanode exclusion list to be populated with the datanode the test just shutdown. But it is not doing so because I modified `watchForCommit` to NOT actually do the Ratis request (remember? the goal of this PR is to 
3. [`TestOzoneClientRetriesOnExceptions.testMaxRetriesByOzoneClient`](https://github.com/apache/ozone/actions/runs/8907392352/job/24479780977?pr=6014#step:6:2212) -- `AssertionFailedError: Expected java.io.IOException to be thrown, but nothing was thrown` looks to be the same type of issue as `TestCommitWatcher#testReleaseBuffersOnException` that I fixed already in 5c68ead09ca3c6235c998c338e9a64db957479fc. Should be an easy one.

smengcl commented:
For 2: `XceiverClientRatis#watchForCommit` had this [logic](https://github.com/apache/ozone/blob/1057c62ec4fe5c68b6ffb3dc6108375b26092e3f/hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientRatis.java#L277-L278) where it initially attempts `ALL_COMMITTED`, but when it failed to do a 3-way commit, it falls back to `MAJORITY_COMMITTED`, which is where it adds the failed datanode to the list.

`XceiverClientRatis#sendCommandAsync` might have to do exactly the same to keep the same behavior now that `watchForCommit` Ratis request is omitted. (Or is it worth keeping such behavior?)

szetszwo commented:
@smengcl , Let's try the snapshot right at the commit of RATIS-1994.  Pushed `3.1.0-5e8ef17-SNAPSHOT`.
```java
commit 5e8ef17e7a17042c99875d398c9e9b9cee8a3ac4
Author: Tsz-Wo Nicholas Sze <szetszwo@apache.org>
Date:   Thu Jan 11 20:38:27 2024 -0800

    RATIS-1994. AsyncApi send() to support optional replication level. (#1006)
```

szetszwo commented:
Just tried -- `TestRandomKeyGenerator.testKeyLargerThan2GB()` can pass with `3.1.0-5e8ef17-SNAPSHOT`.

smengcl commented:
> Just tried -- `TestRandomKeyGenerator.testKeyLargerThan2GB()` can pass with `3.1.0-5e8ef17-SNAPSHOT`.

Thanks so much @szetszwo ! The test passed after rolling back to the point [RATIS-1994](https://github.com/apache/ratis/pull/1006) is just merged.

Do we know what exactly caused the regression? Is it [RATIS-2094](https://issues.apache.org/jira/browse/RATIS-2094)?

smengcl commented:
For transparency this is what I plan to do further with this PR in order to resolve the `testDatanodeExclusionWithMajorityCommit` failure:

1. Pull in [RATIS-2102](https://issues.apache.org/jira/browse/RATIS-2102), which should fix the bugs in [RATIS-1994](https://issues.apache.org/jira/browse/RATIS-1994)
2. Pull in [RATIS-2089](https://issues.apache.org/jira/browse/RATIS-2089), which will allow `NotReplicatedException` to carry `CommitInfoProto` of all nodes in the Raft quorum returned from Raft server (leader) despite of that `NotReplicatedException` (in which case `reply` would be `null` and won't have the `CommitInfoProto`s we need to populate the `failedServers` dead node list)
3. Tweak `sendRequestAsync()` in order to "fake" a reply using the `CommitInfoProto` that is now being passed with RATIS-2089 even in the event of `NotReplicatedException` (which means server `watch()` timed out after trying to wait for all followers to commit to the desired log index).
4. Because `addDatanodetoReply()` has (unfortunately) been used for two different purposes, and I would need to differentiate the reason nodes are added to the list in order to achieve my goal here. (One for adding dead/timed out datanodes to the `failedServers`. Another for adding replier/leader datanode.) As a result, I would need to tweak `addDatanodetoReply()` and `DatanodeDetails` class (if needed) to tag nodes with the reason they are added to the list in order to achieve the purpose.
5. After 3 and 4 are done, tweak `sendCommandAsync()` to call `addDatanodetoReply()` (or a variant of that as I see fit) to add the dead node to the list.
6. Finally after 3,4,5 are done, tweak whoever calls `get()` on `containerCommandResponse` generated by `sendCommandAsync()` in order to keep the same behavior as `BlockOutputStream#watchForCommit()` does before.

jojochuang commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientRatis.java:
```diff
@@ -217,29 +217,46 @@ public ConcurrentMap<UUID, Long> getCommitInfoMap() {
+    final ContainerCommandRequestMessage message =
+        ContainerCommandRequestMessage.toMessage(request, TracingUtil.exportCurrentSpan());
+    if (HddsUtils.isReadOnly(request)) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("sendCommandAsync ReadOnly {}", message);
+      }
```
```suggestion
      LOG.debug("sendCommandAsync ReadOnly {}", message);
```

jojochuang commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientRatis.java:
```diff
@@ -217,29 +217,46 @@ public ConcurrentMap<UUID, Long> getCommitInfoMap() {
+      }
+      return getClient().async().sendReadOnly(message);
+    } else {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("sendCommandAsync write {} {}", writeReplicationLevel, message);
+      }
```
```suggestion
      LOG.debug("sendCommandAsync write {} {}", writeReplicationLevel, message);
```

jojochuang commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java:
```diff
@@ -470,7 +472,8 @@ public static XceiverClientReply writeChunkAsync(
     ContainerCommandRequestProto request = builder.build();
-    return xceiverClient.sendCommandAsync(request);
+//    return xceiverClient.sendCommandAsync(request, ReplicationLevel.MAJORITY_COMMITTED);
+    return xceiverClient.sendCommandAsync(request, ReplicationLevel.ALL_COMMITTED);
```
if WriteChunk is MAJORITY_COMMITTED or ALL_COMMITTED, does it mean waitOnFlushFutures is no longer needed too?

ivandika3 reviewed (commented):
@smengcl Thanks for working on this.

> For transparency this is what I plan to do further with this PR in order to resolve the testDatanodeExclusionWithMajorityCommit failure: ...

The idea looks good. However, we might need reduce the Ratis server watch timeout configuration to trigger NotReplicatedException. 

Additionally, `cluster.shutdownHddsDatanode(datanodes.get(0))` might shutdown the current datanode Ratis leader, which I think will not throw `NotReplicatedException` (e.g. `NotLeaderException` or `GroupMismatchException` (if the pipeline is closed due to `ContainerStateMachine#notifyLeaderChanged` after a new election)). In that case, will the datanode be excluded eventually? I think it might be retried in `KeyOutputStream`  which will exclude it as well, but maybe we can ensure to shutdown the follower datanode instead of leader.

Please let me know what you think.

smengcl commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientRatis.java:
```diff
@@ -363,6 +403,7 @@ public XceiverClientReply sendCommandAsync(
             }
             asyncReply.setLogIndex(reply.getLogIndex());
             addDatanodetoReply(serverId, asyncReply);
+            addDatanodetoReply(serverId, asyncReply);  // add replier node to XceiverClientReply
```
Oops. That was a mistake made during interactive git staging. Thanks for pointing that out. I intended to only append the comment.

smengcl commented:
@ivandika3 

> The idea looks good. However, we might need reduce the Ratis server watch timeout configuration to trigger NotReplicatedException.

Yes indeed, the current Raft server (datanode) watch.timeout is way too high to trigger `NotReplicatedException`. I've tested that as long as I set DN's `raft.server.watch.timeout` lower than client's `raft.client.rpc.request.timeout`, the integration test would trigger the `NotReplicatedException`. So I intend to adjust the timeouts in this PR as well.

> Additionally, `cluster.shutdownHddsDatanode(datanodes.get(0))` might shutdown the current datanode Ratis leader, which I think will not throw `NotReplicatedException` (e.g. `NotLeaderException` or `GroupMismatchException` (if the pipeline is closed due to `ContainerStateMachine#notifyLeaderChanged` after a new election)). In that case, will the datanode be excluded eventually? I think it might be retried in `KeyOutputStream` which will exclude it as well, but maybe we can ensure to shutdown the follower datanode instead of leader.

Good idea. Let me check.

jojochuang commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java:
```diff
@@ -511,22 +514,33 @@ ContainerCommandResponseProto> executePutBlock(boolean close,
         } catch (IOException sce) {
           throw new CompletionException(sce);
         }
+        boolean timedOut = e.getMessage().equals("TIMEOUT");
```
hacky. Is there a better way to detect timeout?

smengcl replied:
This is a workaround to avoid further refactoring. I just need to get the message across without breaking other parts.

smengcl replied:
A better way to do this might be to add a new field to the proto message. I will consider this later.

jojochuang commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java:
```diff
@@ -808,22 +823,33 @@ CompletableFuture<ContainerCommandResponseProto> writeChunkToContainer(
           handleSuccessfulPutBlock(e.getWriteChunk().getCommittedBlockLength(),
               asyncReply, flushPos, byteBufferList);
         }
+        if (timedOut) {
```
extract this area of code so executePutBlock can reuse.

jojochuang commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/storage/ContainerProtocolCalls.java:
```diff
@@ -291,7 +292,10 @@ public static XceiverClientReply putBlockAsync(XceiverClientSpi xceiverClient,
-    return xceiverClient.sendCommandAsync(request);
+    // TODO: Make only putBlock triggered from hsync ALL_COMMITTED.
+    //  While keeping the ones triggered from write() and others at MAJORITY
+    //  and call regular watchForCommit in order to reduce latency for write()s.
```
consider the case where PutBlock is piggybacked in a WriteChunk too.

smengcl replied:
Yes this is considered. Currently a TODO just down below in `writeChunkAsync()`.

When PutBlock is piggybacked in WriteChunk, `putBlockAsync()` won't be called.

smengcl commented:
Provided that all the tests passed. I would still need to address the performance issue in write()s as a result of switching all send()s to ALL_COMMITTED (high latency). I have added TODOs in the code.

jojochuang commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientRatis.java:
```diff
@@ -218,29 +226,69 @@ public ConcurrentMap<UUID, Long> getCommitInfoMap() {
+        } else {
+          throw new CompletionException("Unexpected exception", e);
         }
+        RaftGroupMemberId raftId = RaftGroupMemberId.valueOf(RaftPeerId.valueOf("peer"), RaftGroupId.emptyGroupId());
```
Does "peer" carry any special meaning? I would have thought it would be a UUID string.
It looks to me RaftPeerId.valueOf("peer") is only used in test code or stubs.

jojochuang commented:
The failure in TestHsync is inherent in the test code. The PR just made it more likely to fail.

The WriteChunk/PutBlock waits for ALL_COMMITTED. However, the fact that all peers have committed does not imply they are applied. In other words, the read after write is not linearizable.

I see there are two solutions
(1) enable Ratis linearizable read, and use XceiverClientRatis to read data (unsupported today)
(2) Add a workaround at Ozone client side to wait and retry.

ivandika3 commented:
> enable Ratis linearizable read, and use XceiverClientRatis to read data

Just want to point out that from some benchmarks on OM linearizable read https://github.com/apache/ozone/pull/5288, there are some significant performance hits. This is because there are additional ReadIndex mechanism between follower and reader, among other things.

smengcl commented:
I brought test failures back down to the last one of them: [`TestOzoneClientRetriesOnExceptions.testMaxRetriesByOzoneClient`](https://github.com/apache/ozone/actions/runs/9494639197/job/26165931945?pr=6014#step:5:2287).

`testMaxRetriesByOzoneClient` [expects](https://github.com/apache/ozone/blob/e93e781ffc991571cbd335e338f4d5a8fbe4a223/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneClientRetriesOnExceptions.java#L235-L246) `key.write()` to throw IOException after maximum retries after all the containers of the key is closed. The expected IOException message looks like this (got this on HDDS-7593 branch where the test is passing):

```
java.io.IOException: java.io.IOException: java.io.IOException: java.io.IOException: java.io.IOException: Retry request failed. retries get failed due to exceeded maximum allowed retries number: 3
```

(Looks like the `IOException` was wrapped 5 times.)

While in this PR, the same test is failing because it is no longer throwing `IOException` with the same `key.write()`. I suspect watchForCommit() was responsible for composing the `IOException` from `ContainerNotOpenException`s after a number of retries. But now that watchForCommit() is (mostly) gone, this code path is no longer active and causes this change in behavior. (`ContainerNotOpenException` is still thrown though.)

szetszwo commented:
> ...  it is no longer throwing IOException with the same key.write(). ...

Did the write fail?  We may consider updating the test.

ivandika3 commented on hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientRatis.java:
```diff
@@ -218,29 +226,71 @@ public ConcurrentMap<UUID, Long> getCommitInfoMap() {
+        final long maxCommitIndex;
+        if (e instanceof CompletionException) {
+          final CompletionException ce = (CompletionException) e;
+          if (e.getCause() instanceof NotReplicatedException) {
+            // Unwrap exception to get commitInfos if NotReplicatedException is thrown
+            final NotReplicatedException nre = (NotReplicatedException) ce.getCause();
+            commitInfos = nre.getCommitInfos();
+            maxCommitIndex = commitInfos.stream()
+                .map(CommitInfoProto::getCommitIndex)
+                .max(Long::compareTo).orElse(-1L);
```
Just wondering whether the `NotReplicatedException` is always the first cause in `CompletionException`?

Maybe we need to use `HddsClientUtils#containsException` similar to how `GroupMismatchException` is handled in HDDS-9826?

smengcl commented:
> > ...  it is no longer throwing IOException with the same key.write(). ...
> 
> Did the write fail? We may consider updating the test.

Yes the write failed. Only the write() before key blocks' containers' closure succeeded, verified using `validateData(keyName, data1)`.

smengcl commented:
After digging for a few more hours I realized the problem might be more than just exception handling.

Because it looks like the first `write()` call (before container close) has led to an incorrect internal state on my PR branch:

https://github.com/smengcl/hadoop-ozone/blob/0679fbf84683b3a565468572b8ba28875e2febab/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestOzoneClientRetriesOnExceptions.java#L225

Thus, if I simply comment out the first write(), the second write() (AFTER container closure) will correctly throw `IOException: Retry request failed` that it is expecting.

I suspect the `commitIndexMap` update handling (`updateCommitInfoMap`) may be buggy in this case.

smengcl commented:
Ah. The test `testMaxRetriesByOzoneClient` passed locally after I bump Ratis to 3.1.0. So it might be another Ratis bug (that is fixed in 3.1.0 release).

jojochuang commented:
This PR won't make it into the HDDS-7593 feature branch before the branch gets merged next week.
Please rebase onto master branch after the branch merge. Thanks!

adoroszlai commented:
/pending conflicts
